torch/csrc/jit/codegen/cuda/lower_allocation.cpp

```diff
@@ -0,0 +1,365 @@
+#include <torch/csrc/jit/codegen/cuda/lower_allocation.h>
+#include <torch/csrc/jit/codegen/cuda/dispatch.h>
+#include <torch/csrc/jit/codegen/cuda/instrumentation.h>
+#include <torch/csrc/jit/codegen/cuda/kernel_expr_evaluator.h>
+#include <torch/csrc/jit/codegen/cuda/kernel_ir.h>
+#include <torch/csrc/jit/codegen/cuda/kernel_ir_builder.h>
+#include <torch/csrc/jit/codegen/cuda/kernel_ir_printer.h>
+#include <torch/csrc/jit/codegen/cuda/lower2device.h>
+
+#include <unordered_set>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+namespace {
+
+class AllocationInserter : public kir::IrVisitor {
+ private:
+  struct AllocationInformation {
+    // The for loop that the allocation must be placed in, nullptr if not within
+    // a loop
+    kir::ForLoop* for_loop = nullptr;
+
+    // The expression that this allocation must be placed before
+    kir::Expr* place_before = nullptr;
+
+    // The buffer this allocation is for
+    kir::TensorView* buffer = nullptr;
+
+    // The allocation expression
+    kir::Allocate* alloc_expr = nullptr;
+
+    // Initialization
+    kir::Expr* init_expr = nullptr;
+  };
+
+  // Generate allocation and initialization expressions and place it in info. If
+  // used for a reduction init_val should be initial value, if not a reduction
+  // init_val should be nullptr.
+  void fillAllocInfo(
```

Comment:
why not return AllocationInformation? Unless there's a strong reason for out parameters I'd recommend defaulting to return values (compose much better and avoid confusion about out vs. in/out)